base_report_to_printer/models/ir_actions_report.py

```diff
@@ -0,0 +1,138 @@
+    _inherit = 'ir.actions.report'
+
+    property_printing_action_id = fields.Many2one(
+        comodel_name='printing.action',
+        string='Default Behaviour',
+        company_dependent=True,
+    )
+    printing_printer_id = fields.Many2one(
+        comodel_name='printing.printer',
+        string='Default Printer'
+    )
+    printing_action_ids = fields.One2many(
+        comodel_name='printing.report.xml.action',
+        inverse_name='report_id',
+        string='Actions',
+        help='This field allows configuring action and printer on a per '
+             'user basis'
+    )
+
+    @api.model
+    def print_action_for_report_name(self, report_name):
+        """ Returns if the action is a direct print or pdf
+
+        Called from js
+        """
+        report = self._get_report_from_name(report_name)
+        if not report:
+            return {}
+        result = report.behaviour()[report]
+        serializable_result = {
+            'action': result['action'],
+            'printer_name': result['printer'].name,
+        }
+        return serializable_result
+
+    @api.multi
+    def behaviour(self):
+        result = {}
+        printer_obj = self.env['printing.printer']
+        printing_act_obj = self.env['printing.report.xml.action']
+        # Set hardcoded default action
+        default_action = 'client'
+        # Retrieve system wide printer
+        default_printer = printer_obj.get_default()
+
+        # Retrieve user default values
+        user = self.env.user
+        if user.printing_action:
+            default_action = user.printing_action
+        if user.printing_printer_id:
+            default_printer = user.printing_printer_id
+
+        for report in self:
+            action = default_action
+            printer = default_printer
+
+            # Retrieve report default values
+            report_action = report.property_printing_action_id
+            if report_action and report_action.action_type != 'user_default':
+                action = report_action.action_type
+            if report.printing_printer_id:
+                printer = report.printing_printer_id
+
+            # Retrieve report-user specific values
+            print_action = printing_act_obj.search([
+                ('report_id', '=', report.id),
+                ('user_id', '=', self.env.uid),
+                ('action', '!=', 'user_default'),
+            ], limit=1)
+            if print_action:
+                user_action = print_action.behaviour()
+                action = user_action['action']
+                if user_action['printer']:
+                    printer = user_action['printer']
+
+            result[report] = {
+                'action': action,
+                'printer': printer,
+            }
+        return result
+
+    @api.multi
+    def print_document(self, record_ids, data=None):
+        """ Print a document, do not return the document file """
+        document = self.with_context(
```

Comment:
@sylvain-garancher I'm becoming a bit anti these kind of calls for tightly coupled behaviours.  It makes inheritance difficult as you can't pass a context down a chain of calls.  In this example we could be passing anything from languages, to keys used in the report.  In any case I see no reason not to send the passed context as well.  I'd rather we had

```
ctx = dict(self._env.context, must_skip_send_to_printer=True)
document = self.with_context(**ctx)...
```

Reply:
I'm not sure to understand your concern, but I home this answers:
```
>>> self.env.context
{'tz': False, 'lang': 'en_US'}
>>> self.with_context(must_skip_send_to_printer=True).env.context
{'tz': False, 'lang': 'en_US', 'must_skip_send_to_printer': True}
```

Reply:
Hi

@sylvain-garancher try again but this time with a different context other than default. 

Your example is not adding the existing context, it is merely recreating a default context in the new one.  So any context passed is lost when sending to printer because we are creating a new environment.

So 
```python
self.with_context(example='example').print_document(...)
example won't be passed to the new context.
```

Reply:
@sylvain-garancher weird thing is I tested this for ages and it didn't work.  I just tested now in shell and it did work.  I'll test again properly later, ignore for now.

Reply:
@sylvain-garancher you are probably right.  I can't fully remember now, that was late nights back then, but it was adding arguments that were returned i a dictionary format although I thought I'd **ed the return, but I just tested it and it works.

In truth its great for me I am wrong.  Just allowed me to remove a ton of useless override.